sys/aarch64/boot.c

```diff
@@ -199,6 +207,42 @@ __boot_text static paddr_t build_page_table(void) {
 
   l0[L0_INDEX(DMAP_BASE)] = (pde_t)l1d | L0_TABLE;
 
+#if KASAN /* Prepare KASAN shadow mappings */
+  size_t kasan_sanitized_size =
+    SUPERPAGESIZE + roundup2(va - KASAN_MD_SANITIZED_START,
+                             SUPERPAGESIZE * KASAN_SHADOW_SCALE_SIZE);
+  size_t kasan_shadow_size = kasan_sanitized_size / KASAN_SHADOW_SCALE_SIZE;
+  vaddr_t kasan_shadow_end = KASAN_MD_SHADOW_START + kasan_shadow_size;
+  va = KASAN_MD_SHADOW_START;
+  /* XXX _kasan_sanitized_end is at a high address which is not mapped yet,
+   * so we access it using its physical address instead. */
+  *(vaddr_t *)AARCH64_PHYSADDR(&_kasan_sanitized_end) =
+    KASAN_MD_SANITIZED_START + kasan_sanitized_size;
+  /* Allocate physical memory for shadow area */
+  pa = (paddr_t)bootmem_alloc(kasan_shadow_size);
+
+  while (va < kasan_shadow_end) {
+    if (l0[L0_INDEX(va)] == 0)
+      l0[L0_INDEX(va)] = (pde_t)bootmem_alloc(PAGESIZE) | L0_TABLE;
+
+    pde_t *l1k = (pde_t *)PTE_FRAME_ADDR(l0[L0_INDEX(va)]);
+    if (l1k[L1_INDEX(va)] == 0)
+      l1k[L1_INDEX(va)] = (pde_t)bootmem_alloc(PAGESIZE) | L1_TABLE;
+
+    pde_t *l2k = (pde_t *)PTE_FRAME_ADDR(l1k[L1_INDEX(va)]);
+    if (l2k[L2_INDEX(va)] == 0)
+      l2k[L2_INDEX(va)] = (pde_t)bootmem_alloc(PAGESIZE) | L2_TABLE;
+
+    pde_t *l3k = (pde_t *)PTE_FRAME_ADDR(l2k[L2_INDEX(va)]);
+
+    for (int j = 0; va < kasan_shadow_end && j < PT_ENTRIES; j++) {
+      l3k[L3_INDEX(va)] = pa | ATTR_AP_RW | ATTR_XN | pte_default;
+      va += PAGESIZE;
+      pa += PAGESIZE;
+    }
+  }
+#endif /* !KASAN */
```

Comment:
I believe the convention is for the comment to state the condition of the corresponding `#if`, not its negation.
```suggestion
#endif /* KASAN */
```

Reply:
I'm not sure.  It looks like we don't stick to any convention in `.c` files but I agree with you and I will commit your suggestion.

```
kern/vm_physmem.c
71:#endif

kern/mutex.c
22:#endif
35:#endif
77:#endif

kern/malloc.c
19:#endif /* !KASAN */
102:#endif
369:#endif
415:#endif /* !KASAN */

kern/vmem.c
205:#endif

kern/main.c
90:#endif

kern/Makefile
89:endif

kern/lockdep.c
284:#endif

kern/pool.c
25:#endif
45:#endif
83:#endif /* !KASAN */
219:#endif /* !KASAN */
267:#endif

mips/boot.c
124:#endif /* !KASAN */
141:#endif /* !KASAN */

aarch64/boot.c
55:#endif

aarch64/evec.S
66:.endif
80:.endif
84:.endif
110:.endif

tests/utest.c
132:#endif
141:#endif
149:#endif
166:#endif

tests/vm_map.c
14:#endif
18:#endif

tests/sched.c
43:#endif
97:#endif

tests/pool.c
39:#endif

libkern/stdio/scanf.c
404:#endif
```